src/js/d3ViewComponent.js

```diff
@@ -114,11 +139,103 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    floe.d3ViewComponent.addElementIdToDataKey = function (d3Key, idToAdd, that) {
+        var keyPath = "dataKeys." + d3Key;
+        var elementIds = fluid.get(that.model, keyPath);
+        // Create an empty array at the path if the key isn't currently defined
+        if (elementIds === undefined) {
+            that.applier.change(keyPath, []);
+            elementIds = fluid.get(that.model, keyPath);
+        }
+        if (!elementIds.includes(idToAdd)) {
+            elementIds.push(idToAdd);
+        }
+        that.applier.change(keyPath, elementIds);
+    };
+
+    // Corresponding "remove" functionality to addElementIdToDataKey
+    floe.d3ViewComponent.removeElementIdFromDataKey = function (d3Key, idToRemove, that) {
+        var keyPath = "dataKeys." + d3Key;
+        var elementIds = fluid.get(that.model, keyPath);
+        fluid.remove_if(elementIds, function (currentId) {
+            return currentId === idToRemove;
+        });
+        that.applier.change(keyPath, elementIds);
+    };
+
+    // Given an array "elements" consisting of element IDs, returns a joined
+    // string of IDs suitable for use as a jQuery selector to select all
+    // those IDs
+    floe.d3ViewComponent.getElementIdsAsSelector = function (elementIds) {
+        if (fluid.isArrayable(elementIds)) {
+            var elemIdCollectionWithPreface = fluid.transform(elementIds, function (elemId) {
+                    return "#" + elemId;
+                });
+            var keyedElements = elemIdCollectionWithPreface.join(", ");
+            return keyedElements;
+        }
+    };
+
+    // Given an array of D3 data keys, returns all affiliated D3-bound elements
+    // using the model's dataKeys information
+    floe.d3ViewComponent.getElementsByDataKeys = function (dataKeys, that) {
+        var matchedElements = [];
+        fluid.each(dataKeys, function (dataKey) {
+            var elementIds = that.model.dataKeys[dataKey];
+
+            var selector = floe.d3ViewComponent.getElementIdsAsSelector(elementIds);
+            matchedElements.push(selector);
+        });
+        return $(matchedElements.join(","));
+    };
+
+    // Given a D3 data key, returns all D3-bound elements that aren't associated
+    // with that key
+
+    floe.d3ViewComponent.getElementsNotMatchingDataKey = function (dataKey, that) {
+        var dataKeys = fluid.copy(that.model.dataKeys);
+        fluid.remove_if(dataKeys, function (currentObj, currentKey) {
+            return currentKey === dataKey;
+        });
+        return floe.d3ViewComponent.getElementsByDataKeys(Object.keys(dataKeys), that);
+    };
+
+    // Given a selection of D3 elements, an ID and a CSS class, turns that
+    // class on for any elements matching the ID and makes sure it's turn off
+    // for any elements not matching it
+    floe.d3ViewComponent.toggleCSSClassByDataId = function (id, toggleClass, that) {
+        var associatedElements = that.getElementsByDataKey(id);
```

Comment:
This function finds the matched ones by id first, then iterate thru ```that.model.dataKeys``` to remove all matched ones then iterate thru the rest elements. I'm thinking it might be a bit more efficient to iterate thru ```that.model.dataKeys``` once and toggle the class for all elements.